src/NServiceBus.Core/Sagas/SagaMetadata.cs

```diff
@@ -110,5 +119,273 @@ public bool TryGetFinder(string messageType, out SagaFinderDefinition finderDefi
+                {
+                    var args = interfaceType.GetGenericArguments();
+                    //since we dont want to process the IFinder type
+                    if (args.Length != 2)
+                    {
+                        continue;
+                    }
+
+                    var entityType = args[0];
+                    if (entityType != sagaEntityType)
+                    {
+                        continue;
+                    }
+
+                    var messageType = args[1];
+                    if (!conventions.IsMessageType(messageType))
+                    {
+                        var error = string.Format("A custom IFindSagas must target a valid message type as defined by the message conventions. Please change '{0}' to a valid message type or add it to the message conventions. Finder name '{1}'.", messageType.FullName, finderType.FullName);
+                        throw new Exception(error);
+                    }
+
+                    var existingMapping = mapper.Mappings.SingleOrDefault(m => m.MessageType == messageType);
+                    if (existingMapping != null)
+                    {
+                        var bothMappingAndFinder = string.Format("A custom IFindSagas and an existing mapping where found for message '{0}'. Please either remove the message mapping for remove the finder. Finder name '{1}'.", messageType.FullName, finderType.FullName);
+                        throw new Exception(bothMappingAndFinder);
+                    }
+                    mapper.ConfigureCustomFinder(finderType, messageType);
+                }
+            }
+        }
+
+        static void SetFinderForMessage(SagaToMessageMap mapping, Type sagaEntityType, List<SagaFinderDefinition> finders)
+        {
+            if (mapping.HasCustomFinderMap)
+            {
+                finders.Add(new SagaFinderDefinition(typeof(CustomFinderAdapter<,>).MakeGenericType(sagaEntityType, mapping.MessageType), mapping.MessageType.FullName, new Dictionary<string, object>
+                {
+                    {"custom-finder-clr-type", mapping.CustomFinderType}
+                }));
+
+            }
+            else
+            {
+                finders.Add(new SagaFinderDefinition(typeof(PropertySagaFinder<>).MakeGenericType(sagaEntityType), mapping.MessageType.FullName, new Dictionary<string, object>
+                {
+                    {"property-accessor", mapping.MessageProp},
+                    {"saga-property-name", mapping.SagaPropName}
+                }));
+            }
+        }
+
+        static IEnumerable<SagaMessage> GetAssociatedMessages(Type sagaType)
+        {
+            var result = GetMessagesCorrespondingToFilterOnSaga(sagaType, typeof(IAmStartedByMessages<>))
+                .Select(t => new SagaMessage(t.FullName, true)).ToList();
+
+            foreach (var messageType in GetMessagesCorrespondingToFilterOnSaga(sagaType, typeof(IHandleMessages<>)))
```

Comment:
Looks like some repetition here could be removed (not sure). These two loops look similar.